azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py

```diff
@@ -322,3 +404,24 @@ def send_event(self, message, callback=None):
 
     def _on_shared_access_string_updated(self):
         self._trig_on_shared_access_string_updated()
+
+    def enable_input_messages(self, callback=None):
+        self._trig_enable_receive(callback)
+        # TODO : Discuss state machine configuration with team
+
+    def enable_c2d_messages(self, callback=None, custom_handler=None):
+        self._subscribe_callback = callback
+        self._trig_enable_receive(callback, custom_handler)
+        # TODO : Discuss state machine configuration with team
+
+    def _call_subscribe(self, event_data):
+        _receive_topic = self._get_receive_topic()
+        logger.info("c2d receive message topic is " + _receive_topic)
+        self._mqtt_provider.subscribe(_receive_topic, qos=1)
+        custom_handler = event_data.args[1]
+        if custom_handler:
+            self._mqtt_provider.on_mqtt_message_received = custom_handler
+        else:
+            self._mqtt_provider.on_mqtt_message_received = (
+                self._on_provider_message_received_callback
+            )
```

Comment:
Here is another place where you assume that C2D and input messages are mutually exclusive based on something that is not particularly explicit.

Reply:
this was only done when a custom handler was passed...its done a bit differently now..

---
In reply to: [246549489](https://github.com/Azure/azure-iot-sdk-python-preview/pull/36#discussion_r246549489) [](ancestors = 246549489)